### ISSUE TITLE
XPathUnhandledException on export of forms using 3rd party client functions

### ISSUE THREAD

**Issue by joeflack4** [opendatakit/briefcase#796]
#### Software versions 
- Briefcase v1.16.3
- Java v1.8.x
- Windows 10

#### Problem description
When doing an export, `XPathUnhandledException` occurs, causing Briefcase to hang indefinitely.

#### Steps to reproduce the problem
1. **XLSForm**: Convert to XML an XLSForm that has a custom functionality intended to be used by a 3rd party client, i.e. not ODK Collect, e.g. [JHU Collect](https://github.com/PMA-2020/collect).
2. **Aggregate**: Upload form def to aggregate
3. **Client**: Download, complete, and submit form using a 3rd party client
4. **Briefcase**: Load GUI via the CLI
5. **Briefcase**: Fetch submissions
6. **Briefcase**: Attempt to export
7. **Briefcase**: Observe error stacktrace in terminal window

#### Expected behavior
- When exporting, should not error / freeze / crash.
- Briefcase should only care about the submission data; should not care about implementation details of how that data was created.

#### Other information 
**The particular issue**
The JHU Collect feature causing this issue is `trim()`, which clears extraneous whitespace. I believe that ODK Collect now supports this in a somewhat new feature, `normalize-space()`. In any event, I suppose it doesn't matter the specific 3rd party feature that was used. I'm just thinking that it would be ideal if Briefcase did not fully parse / evaluate / validate the form definition during the export (or any) process, unless there is a need for it that I am not seeing.

**Stack trace**
```java
C:\Users\annro>java -jar Downloads\ODK-Briefcase-v1.16.3.jar
Exception in thread "Thread-4" java.lang.RuntimeException: Error evaluating field 'firstname': The problem was located in calculate expression for /FSQ/firstname
XPath evaluation: cannot handle function 'trim'
       at org.javarosa.core.model.LatestDagBase.evaluateTriggerable(LatestDagBase.java:139)
       at org.javarosa.core.model.LatestDagBase.doEvaluateTriggerables(LatestDagBase.java:63)
       at org.javarosa.core.model.Safe2014DagImpl.evaluateTriggerables(Safe2014DagImpl.java:366)
       at org.javarosa.core.model.Safe2014DagImpl.initializeTriggerables(Safe2014DagImpl.java:452)
       at org.javarosa.core.model.Safe2014DagImpl.initializeTriggerables(Safe2014DagImpl.java:430)
       at org.javarosa.core.model.FormDef.initializeTriggerables(FormDef.java:721)
       at org.javarosa.core.model.FormDef.initialize(FormDef.java:1306)
       at org.opendatakit.briefcase.export.FormDefinition.getFormControls(FormDefinition.java:148)
       at org.opendatakit.briefcase.export.FormDefinition.from(FormDefinition.java:142)
       at org.opendatakit.briefcase.ui.export.ExportPanel.lambda$export$10(ExportPanel.java:190)
       at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
       at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source)
       at java.util.stream.AbstractPipeline.copyInto(Unknown Source)
       at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
       at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(Unknown Source)
       at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(Unknown Source)
       at java.util.stream.AbstractPipeline.evaluate(Unknown Source)
       at java.util.stream.ReferencePipeline.forEach(Unknown Source)
       at org.opendatakit.briefcase.reused.job.JobsRunner.launchAsync(JobsRunner.java:63)
       at org.opendatakit.briefcase.reused.job.JobsRunner.launchAsync(JobsRunner.java:51)
       at org.opendatakit.briefcase.ui.export.ExportPanel.export(ExportPanel.java:217)
       at java.lang.Thread.run(Unknown Source)
Caused by: org.javarosa.xpath.XPathUnhandledException: The problem was located in calculate expression for /FSQ/firstname
XPath evaluation: cannot handle function 'trim'
       at org.javarosa.xpath.expr.XPathFuncExpr.eval(XPathFuncExpr.java:486)
       at org.javarosa.xpath.XPathConditional.evalRaw(XPathConditional.java:71)
       at org.javarosa.core.model.condition.Triggerable.evalRaw(Triggerable.java:219)
       at org.javarosa.core.model.condition.Recalculate.eval(Recalculate.java:63)
       at org.javarosa.core.model.condition.Triggerable.apply(Triggerable.java:152)
       at org.javarosa.core.model.LatestDagBase.evaluateTriggerable(LatestDagBase.java:125)
       ... 21 more
```

**Versions confirmed to work / not work w/ 3rd party xpath functions**
Working: 1.9.x
Not working: 1.16.3, 1.15.x

**Comment by ggalmazor** [comment]
Hi, @joeflack4! It would help us a lot if we could get example forms and submissions to reproduce and fix the bug. Could you maybe upload them to the sandbox server, or zip the files and attach them to the issue, please?

In the meantime I'll try to work with the stacktrace you've attached

**Comment by ggalmazor** [comment]
Forms with custom functions like `trim` will work in any version of Briefcase previous to the introduction of the "split select multiple fields" feature, which requires form initialization.

Form initialization in JavaRosa requires declaring handlers for custom functions beforehand, but Briefcase doesn't scan for those before doing so.

**Comment by ggalmazor** [comment]
We could compile a list of functions inspecting the `triggerableDAG`list inside `FormDef.dagImpl` and then filter it to get a list of custom functions, but that would require changes in JR. Thoughts, @lognaturel?

**Comment by ggalmazor** [comment]
I've tried a brute force approach:
![image](https://user-images.githubusercontent.com/205913/64236960-b4fb6c00-cefb-11e9-8733-0b8e0d25f7cd.png)

This code triggers any triggerable defined for model fields and inspects the exception message to get a list of missing function handlers. Then, it creates dummy `IFunctionHandler`s (already was doing that for `pulldata`) and registers them.

Suepr hacky, but it works. How does it feel?

**Comment by ggalmazor** [comment]
By the way, I'm reproducing the issue with this form:
```xml
<?xml version="1.0"?>
<h:html xmlns="http://www.w3.org/2002/xforms" xmlns:h="http://www.w3.org/1999/xhtml">
  <h:head>
    <h:title>Custom function form</h:title>
    <model>
      <instance>
        <data id="custom-function-form">
          <some-field/>
          <some-field-trimmed/>
          <meta>
            <instanceID/>
          </meta>
        </data>
      </instance>
      <bind nodeset="/data/some-field" type="string"/>
      <bind nodeset="/data/some-field-trimmed" type="string" calculate="trim(/data/some-field)"/>
      <bind calculate="concat('uuid:', uuid())" nodeset="/data/meta/instanceID" readonly="true()" type="string"/>
    </model>
  </h:head>
  <h:body>
    <input ref="/data/some-field">
      <label>Some field</label>
    </input>
  </h:body>
</h:html>
```

**Comment by ggalmazor** [comment]
I've created a draft PR with the code I've tried at #797

**Comment by joeflack4** [comment]
@ggalmazo Hey Guillermo. Big, big kudos from the PMA team for addressing this, and so quickly. I anticipated this would sit here for a long while.

Apologies for not seeing your comments until now, and that I did not include attachments. I'm glad you were able to reproduce.

Indeed, I experienced the same when using JavaRosa to initialize formDefs in [XformTest](http://github.com/pma-2020/xform-test), and I figured that this was the same issue in newer versions of Briefcase. I do agree with the suggested update to JavaRosa. There's probably a better way as well; better than a static list of custom functions. But I agree w/ the approach to the quick fix to Briefcase (thanks for #797!); and a similar quick fix to JavaRosa. An new GitHub issue can always be posted for a more optimal solution in tandem with the quick fix. Would you like me to post an issue with the suggested change in JavaRosa?

**Comment by lognaturel** [comment]
As you say, parsing the exception messages really doesn't feel great, @ggalmazor. Have you tried your proposed solution with a form that has a lot of dependent logic? My guess is that it results in a noticeable slowdown.

> but that would require changes in JR

I think this would be appropriate since other clients would benefit as well. It would be purely additive so virtually no-risk. I can think of two designs: adding a method that returns a list of strings representing the names of functions without handlers or adding a boolean variant to one of the parsing methods to ignore functions without handlers (by providing no-op handlers at the JR level). Is there a reason you're reluctant to go straight to that?

**Comment by ggalmazor** [comment]
>  Is there a reason you're reluctant to go straight to that?

Not at all! I was just trying to provide a quick&dirty solution while we discussed changes to JR. I agree that it's the best course of action. A method that returns a list of functions without handlers looks like the best option.

If we get that going, we can ship the fix in v1.17.1

**Comment by lognaturel** [comment]
https://github.com/opendatakit/javarosa/pull/483 is broader and riskier than I had anticipated. I don't think it's realistic to get into v1.17.1. If you want to get a patch for custom functions then, maybe you're right that a quick and dirty solution might be a good first step.

If the idea is to replace it shortly and you can verify that performance is not too heavily affected, maybe that's reasonable. In the mean time, we can work on the JR side.
